### PR TITLE
Fix custom natural gas meter issue in tariff

### LINKED ISSUE #7999: Natural gas meter reporting as Other in Energy Cost Summary tariff report
Issue overview
--------------
User was confused with reporting tariffs for natural gas equipment.

image

The use of custom meters confused the tariff cost model.

    Meter:Custom,
      Building Natural Gas,                   !- Name
      NaturalGas,                             !- Fuel Type
      ,                                       !- Key Name 1
      Gas:Facility;                           !- Output Variable or Meter Name 1

    UtilityCost:Tariff,
      Sample with All Utilities_NGas,         !- Name
      Building Natural Gas,                   !- Output Meter Name
      Therm,                                  !- Conversion Factor Choice

If the user had used Gas:Facility as the Output Meter Name the tariff object cost result would show up in the Natural Gas row in the Energy Cost Summary. But since they used the name of a custom meter the cost information shows up in the Other category.

The code that selects where this information is reported is in EconomicTariff.cc.

The first IF here looks at tariff(iTariff).kindElectricMtr >= kindMeterElecSimple, or any type of electricity meter.

The next else if looks only at the Gas:Facility meter which is why this custom meter was reported to the Other category.

The suggestion here is to change the first else if to look at all natural gas meters [e.g., else if (tariff(iTariff).kindGasMtr == kindMeterGas)].

Also, this first variable is not used in this function. And depending on how this is implemented, the second may also not be needed. 

    elecFacilMeter = GetMeterIndex("ELECTRICITY:FACILITY");
    gasFacilMeter = GetMeterIndex("GAS:FACILITY");
    distCoolFacilMeter = GetMeterIndex("DISTRICTCOOLING:FACILITY");
    distHeatFacilMeter = GetMeterIndex("DISTRICTHEATING:FACILITY");

    if (tariff(iTariff).kindElectricMtr >= kindMeterElecSimple) {
        if (tariff(iTariff).totalAnnualEnergy > elecTotalEne) elecTotalEne = tariff(iTariff).totalAnnualEnergy;
        elecTotalCost += tariff(iTariff).totalAnnualCost;
        elecTariffNames += ' ' + tariff(iTariff).tariffName;
        elecUnits = tariff(iTariff).convChoice;
    } else if (tariff(iTariff).reportMeterIndx == gasFacilMeter) {
        if (tariff(iTariff).totalAnnualEnergy > gasTotalEne) gasTotalEne = tariff(iTariff).totalAnnualEnergy;
        gasTotalCost += tariff(iTariff).totalAnnualCost;
        gasTariffNames += ' ' + tariff(iTariff).tariffName;
        gasUnits = tariff(iTariff).convChoice;
        gasDemWindowUnits = tariff(iTariff).demandWindow;
    } else if (tariff(iTariff).reportMeterIndx == distCoolFacilMeter) {
        if (tariff(iTariff).totalAnnualEnergy > distCoolTotalEne) distCoolTotalEne = tariff(iTariff).totalAnnualEnergy;
        distCoolTotalCost += tariff(iTariff).totalAnnualCost;
        distCoolTariffNames += ' ' + tariff(iTariff).tariffName;
        distCoolUnits = tariff(iTariff).convChoice;
        distCoolDemWindowUnits = tariff(iTariff).demandWindow;
    } else if (tariff(iTariff).reportMeterIndx == distHeatFacilMeter) {
        if (tariff(iTariff).totalAnnualEnergy > distHeatTotalEne) distHeatTotalEne = tariff(iTariff).totalAnnualEnergy;
        distHeatTotalCost += tariff(iTariff).totalAnnualCost;
        distHeatTariffNames += ' ' + tariff(iTariff).tariffName;
        distHeatUnits = tariff(iTariff).convChoice;
        distHeatDemWindowUnits = tariff(iTariff).demandWindow;
    } else {
        if (tariff(iTariff).totalAnnualEnergy > otherTotalEne) otherTotalEne = tariff(iTariff).totalAnnualEnergy;
        otherTotalCost += tariff(iTariff).totalAnnualCost;
        othrTariffNames += ' ' + tariff(iTariff).tariffName;
        othrUnits = tariff(iTariff).convChoice;
        othrDemWindowUnits = tariff(iTariff).demandWindow;
    }

To create a defect file, use a tariff example file and add a Meter:Custom natural gas object as shown here and point to it in the tariff object.

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jcyuan2020**
Pull request overview
---------------------
 - Fixes #7999
 - Also, changed the LEED report "Other" category reporting so that water usage and cost is not reported in "Energy" tables. This would expect to cause some diffs for the current example tariff reporting example case(s). 
 
**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/EconomicTariff.cc [hunk 1]
```diff
@@ -4027,8 +4027,8 @@ void LEEDtariffReporting(EnergyPlusData &state)
 
     using namespace OutputReportPredefined;
 
-    int elecFacilMeter;
-    int gasFacilMeter;
+    // int elecFacilMeter;
+    // int gasFacilMeter;
```
Comment (mitchute) [inline src/EnergyPlus/EconomicTariff.cc]: Could you remove these (and similar) unused lines?

**Review by mitchute — CHANGES_REQUESTED** [review]
What's the expected behavior here? I ran the code with the 5ZoneEconomicsTariffAndLifeCycleCosts file, and the before/after Energy Cost Summary tables are below.

Before
Screen Shot 2021-06-10 at 5 50 50 AM

After
Screen Shot 2021-06-10 at 5 48 44 AM

I see that the "other" category is now 0, but was that supposed to move to the natural gas category? Also, the Total Energy Cost columns do not sum properly, both before and after.

File: src/EnergyPlus/EconomicTariff.cc (hunk 1 above)
Comment (jcyuan2020) [reply src/EnergyPlus/EconomicTariff.cc]: @mitchute Thanks! Will do a clean up this. 
For the example file diffs, the example file does not have customized natural gas items at all. The "other" category in the "before" case is actually water, which is not supposed be there--this is actually belongs to the the second part of the fix (se the second listed item in the top comment).

Comment (mitchute) [reply src/EnergyPlus/EconomicTariff.cc]: @jcyuan2020 sorry, I wasn't clear in my original comment. What I actually did was modify that example file to change the "ExampleA-Gas" UtilityCost:Tariff and add the meter, as shown below (which is from the original GH issue).

```
Meter:Custom,
  Building Natural Gas,                   !- Name
  NaturalGas,                             !- Fuel Type
  ,                                       !- Key Name 1
  Gas:Facility;                           !- Output Variable or Meter Name 1

UtilityCost:Tariff,
  Sample with All Utilities_NGas,         !- Name
  Building Natural Gas,                   !- Output Meter Name
  Therm,                                  !- Conversion Factor Choice
```

The tables posted are with those changes applied.

Comment (jcyuan2020) [reply src/EnergyPlus/EconomicTariff.cc]: @mitchute Thanks again! A charge entry needs to be added for the newly added items, like what's been used in the unit test here:
https://github.com/NREL/EnergyPlus/pull/8787/files#diff-2483eb12f59875a8f7d44cc395a891fd9b94b679aa556e6989637af4b22ef02fR1299-R1305
```
UtilityCost:Charge:Simple,
    FlatEnergyCharge-Gas_Custom,  !- Utility Cost Charge Simple Name
    Sample with All Utilities_NGas,  !- Tariff Name
    totalEnergy,             !- Source Variable
    Annual,                  !- Season
    EnergyCharges,           !- Category Variable Name
    0.50;                    !- Cost per Unit Value or Variable Name
```

I would say this is actually indeed a little bit tricky and it took me sometime to figure out when I was developing a suitable test case.

Comment (mitchute) [reply src/EnergyPlus/EconomicTariff.cc]: OK, thanks for pointing that out -- I figured it out. Here's how it looks now.

Before
Screen Shot 2021-06-10 at 10 00 25 AM

After
Screen Shot 2021-06-10 at 9 57 56 AM

Looks good to me. Unit tests run locally, and the table now shows the expected results.

**Review by mitchute — APPROVED** [review]

File: src/EnergyPlus/EconomicTariff.cc (hunk 1 above)
Comment (mitchute) [reply src/EnergyPlus/EconomicTariff.cc]: Although, the total is now different... is that a problem? @jcyuan2020 ?

Comment (jcyuan2020) [reply src/EnergyPlus/EconomicTariff.cc]: The difference of the two totals (4763.24 - 3091.71) equals to 1671.53, which is the water cost (in the original branch this was categorized into "Other"). It looks like a good match and I think it is the expected outcome after excluding water costs (and usages) in this "Energy" report.

Comment (mitchute) [reply src/EnergyPlus/EconomicTariff.cc]: OK, thanks for the clarification. I'll merge this once CI wraps up.

**Comment by mitchute** [comment]
I ran regressions on the files with errors locally on Mac and I'm not seeing any problems. Maybe there's a different code path that gets exercised on CI vs locally for this? Either way, the problem isn't with this work. Merging.

**Pull request merged**